keps/sig-auth/4412-projected-service-account-tokens-for-kubelet-image-credential-providers/README.md

```diff
@@ -0,0 +1,1034 @@
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+To keep this KEP small both in terms of implementation and scope,
+the design is written to meet the following criteria:
+
+- No changes to any Kubernetes REST APIs
+- No changes to the CRI API
+- No changes to how the Kubelet interacts with the CRI API
+- No changes to any registry
+- Minimal changes to Kubelet
+- Minimal changes to Kubelet credential providers
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+TODO(aramase)
+1. add details about the ensure image pulls KEP and impact on this KEP - Talk about how the caching logic will require changes.
+
+#### Credential Provider Configuration
+
+Add `ServiceAccountTokenAudience` and `ServiceAccountAnnotations` field to `CredentialProvider` struct in `CredentialProviderConfig`.
+
+1. `ServiceAccountTokenAudience` is the intended audience for the credentials. If set, the kubelet will generate a service account token for the audience and pass it to the plugin.
+   1. Only a single audience can be configured.
+   2. The credential provider can be set up multiple times for each audience.
+2. `ServiceAccountAnnotations` is the list of annotations on the service account that plugin can use to extract additional information required to fetch credentials. The
+annotations will be passed to the plugin along with the image and service account token. The plugin is responsible for validating and using the annotations as needed.
+   1. Service account annotations are significantly more stable (and map naturally to the service account being the partioning dimension).
+   2. This field makes the provider opt into getting the annotations it wants, avoids sending arbitrary annotation contents down to the plugin (including stuff like client-side apply annotations) and shrinks the set of annotations that could invalidate the cache.
+
+```diff
+--- a/pkg/kubelet/apis/config/types.go
++++ b/pkg/kubelet/apis/config/types.go
+@@ -653,10 +653,19 @@ type CredentialProvider struct {
+        // +optional
+        Env []ExecEnvVar
+ 
++       // ServiceAccountTokenAudience is the intended audience for the projected service account token.
++       // If set, the kubelet will generate a service account token for the audience and pass it to the plugin.
++       // +optional
++       ServiceAccountTokenAudience string
++
++       // ServiceAccountAnnotations is the list of annotations to be passed to the plugin.
```

Comment:
aren't annotations often key-value pairs? are users passing keys to this and then the kubelet populates the value? or is it only keys, or some combination?

Reply:
The `ServiceAccountAnnotations` in the credential provider config is a list indicating which annotations from Kubernetes Service Account (KSA) the kubelet needs to send as part of the credential provider request.

For example,

```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: my-service-account
  namespace: my-namespace
  annotations:
    domain.io/identity-id: "12345"
    domain.io/identity-type: "user"
    domain.io/annotation-that-will-not-be-passed: "value"
```
```yaml
apiVersion: kubelet.config.k8s.io/v1
kind: CredentialProviderConfig
providers:
  - name: acr-credential-provider
    matchImages:
      - "*.registry.io/*"
    defaultCacheDuration: "10m"
    apiVersion: credentialprovider.kubelet.k8s.io/v1
    serviceAccountTokenAudience: my-audience
    serviceAccountAnnotations:
      - domain.io/identity-id
      - domain.io/identity-type
```

Example credential provider request:

```json
{
  "image": "my-image",
  "serviceAccountToken": "<service-account-token>",
  "serviceAccountAnnotations": {
    "domain.io/identity-id": "12345",
    "domain.io/identity-type": "user"
  }
}
```

I have this documented in the KEP as well. I guess the name `ServiceAccountAnnotations` might be misleading. The intent is to say the field here defines what subset of annotations from KSA need to be plumbed to the plugin. I'm open to suggestions on the field name.

Reply:
I think I missd it in the kep generally, maybe it could use a summary in this section so it's clear that's the goal. also maybe `ServiceAccountAnnotationKeys` ?